### PR TITLE
Update NERSC system packages

### PR DESCRIPTION
**Pull request by bcfriesen**
Cray's OS provides many standard packages that Spack does not need to build from source. Adding these to the NERSC `packages.yaml` speeds up builds of many different packages.

### CODE REVIEW

**Comment by bcfriesen** [comment]
Any update on this? Love it? Hate it?

**Comment by tgamblin** [comment]
Maybe we should start a codeowners file for this so that the right people are contacted.

@mamelara: can you review?

One worry is that everything here is

**Comment by tgamblin** [comment]
One major worry here is that setting everything to `buildable: false` will cause any build that depends on an unavailable version of these packages to fail.  

The line for Cmake in particular will cause amrex to fail, as it requires a newer cmake version.

**Comment by bcfriesen** [comment]
> One major worry here is that setting everything to buildable: false will cause any build that depends on an unavailable version of these packages to fail.

Ah OK, I was copy+pasting from another example; did not understand exactly what `buildable: false` meant. Should all instances of that be removed?

**Comment by mpbelhorn** [comment]
@bcfriesen

> Should all instances of that be removed?

That depends on if 
* The package is a build-time dependency
* or a runtime dependency that will need to work on both the frontend and backend nodes
* and what the frontend node and backend node architectures are.

I'm not sure the situation I'm about to describe applies to Cori and Edison, but it's the general case and it does apply to all of the OLCF Crays.

On the OLCF Crays, building with the cray platform (ie, using the Cray compiler wrappers) targets the compute node architecture by default. Unfortunately, in our case, the compute node instruction set is a superset of frontend nodes. So using spack's cray platform produces binaries that cannot be executed on the login nodes which is where building takes place. Therefore on our machines, any package that is needed at build time is marked as unbuildable and we ensure that there is an existing external package listed in packages.yaml (either by path or modulefile) that will satisfy all dependencies. Unfortunately we don't know if a package is missing until a build fails.

If a needed build-time package is not provided by the OS or CrayPE, as in the CMake example that @tgamblin described regarding amrex, we build it with another instance of spack that is configured to use only gcc compilers that target the login nodes using the `arch=linux-sles*-x86_64` platform (ie, in an environment where `unset CRAYPE_VERSION` so spack doesn't think it's on a Cray). These builds by the second spack instance are then added as external builds in the first spack instance's `packages.yaml`. 

We've found that having two spack instances is necessary rather than passing spack very convoluted mixed-compiler or mixed-arch specs so that spack's concretizer doesn't attempt to mix common link dependencies between the package that must run on the build node with packages that must run on the compute nodes. And this trick only works because on all of our machines, the instruction set available on the login nodes is a subset of the compute nodes so builds targeting the login nodes will run on both types of nodes. If a site has a Cray with x86_64 frontend nodes and say aarch64 backend nodes, this scheme wouldn't work.

The ultimate solution to this problem would be that when spack builds a package with `arch=cray-cnl*-*`, spack's concretizer would be configurable to automatically use something like `arch=linux-sles*-x86_64` for build-time-only dependency packages (including their dependents) and not attempt to share link dependencies common to both the original target package and it's build-time dependencies. If that makes any sense. I'm on a mobile phone and can't currently post examples of concretized specs that would more clearly express the issue. I'll try to update this post later with examples.

One work-around is to allow spack to build any package it wants on a cray but to always target `arch=cray-cnl*-<generic>` where `<generic>` is an Cray compiler wrapper architecture target that has a common instruction set on all the nodes of the machine. It's not a good work-around since we generally want applications for the compute nodes to use the full instruction set. In the case of our Crays, the compute nodes have more performant SIMD extensions and floating point processing units that would not be taken advantage of if using this work-around.

**TL;DR**: Basically, until spack's concretizer can be updated to better handle mixed architectures and compilers for build-, runtime-, and link- dependencies, packages that need to run on the frontend nodes of mixed-architecture Crays need to be unbuildable external packages for a spack instance that will use the cray platform to target the backend nodes. Alternatively, sites should either never buy machines with different architectures on the build nodes vs the compute nodes or otherwise prevent compiling from the compute nodes as is the case with many (most? All?) Crays.

**Comment by bcfriesen** [comment]
Hi @mpbelhorn ,

thanks for the detail, this is useful to know. I doubt that changing NERSC's procurement strategy to accommodate shortcomings in spack's concretizer is realistic, so perhaps your other spack workaround is a better solution.

@tgamblin regarding your earlier comment:

> The line for Cmake in particular will cause amrex to fail, as it requires a newer cmake version.

Maybe I am doing something wrong, but the following seems to work for me:

```bash
git clone https://github.com/spack/spack.git --branch develop --single-branch
git clone git@github.com:bcfriesen/spack-configs.git --branch update_NERSC_system_packages --single-branch
. ./spack/share/spack/setup-env.sh
spack -C ./spack-configs/NERSC/cori_and_edison install amrex
```

The output is:

```console
==> cmake@3.5.2 : externally installed in /usr
==> cmake@3.5.2 : generating module file
==> cmake@3.5.2 : registering into DB
==> mpich@7.7.3 : has external module in cray-mpich
==> mpich@7.7.3 : is actually installed in /opt/cray/pe/mpt/7.7.6/gni/mpich-intel/16.0
==> mpich@7.7.3 : generating module file
==> mpich@7.7.3 : registering into DB
==> python@2.7.13 : externally installed in /usr
==> python@2.7.13 : generating module file
==> python@2.7.13 : registering into DB
==> Installing amrex
==> Searching for binary cache of amrex
==> Warning: No Spack mirrors are currently configured
==> No binary for amrex found: installing from source
==> Cloning git repository: https://github.com/AMReX-Codes/amrex.git at commit 260b53169badaa760b91dfc60ea6b2ea3d9ccf06
==> No checksum needed when fetching with git
==> Already staged amrex-18.10.1-gbpoh3t33zyvwgs47hq2i224icozagej in /tmp/spack/var/spack/stage/amrex-18.10.1-gbpoh3t33zyvwgs47hq2i224icozagej
==> No patches needed for amrex
==> Building amrex [CMakePackage]
==> Executing phase: 'cmake'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed amrex
  Fetch: 15.90s.  Build: 1m 51.80s.  Total: 2m 7.70s.
```

If I remove the `buildable: false` lines from all of the system packages in `/usr` that I added in this pull request, the output is identical. So if I remove `buildable: false` from each of those commits, would this PR be more palatable?

**Comment by bcfriesen** [comment]
Any update on this?

File: NERSC/cori_and_edison/packages.yaml [hunk 1]
```diff
@@ -82,3 +82,79 @@ packages:
...
+             xz@5.0.5: /usr
+    readline:
+        buildable: false
+        paths:
+             readline@6.3: /usr
+    zlib:
+        buildable: false
+        paths:
+             zlib@1.2.8: /usr
+    python:
```
Comment (mamelara) [inline NERSC/cori_and_edison/packages.yaml]: I think it might be better to let Spack just build Python.

File: NERSC/cori_and_edison/packages.yaml [hunk 2]
```diff
@@ -82,3 +82,79 @@ packages:
...
+    automake:
+        buildable: false
+        paths:
+             automake@1.13.4: /usr
+    bzip2:
+        buildable: false
+        paths:
+             bzip2@1.0.6: /usr
+    cmake:
+        buildable: false
```
Comment (mamelara) [inline NERSC/cori_and_edison/packages.yaml]: The AMReX team was having issues with CMake since they depend on 3.13.0. We should probably allow building of CMake since this one is old

File: NERSC/cori_and_edison/packages.yaml [hunk 3]
```diff
@@ -82,3 +82,79 @@ packages:
             trilinos@12.12.1.1%gcc: cray-trilinos
             trilinos@12.12.1.1%intel: cray-trilinos
             trilinos@12.12.1.1%cce: cray-trilinos
+    openssl:
+        buildable: false
+        paths:
+             openssl@1.0.2j-fips: /usr
+    autoconf:
+        buildable: false
```
[the remark this answers is not in the document]
Comment (bcfriesen) [reply NERSC/cori_and_edison/packages.yaml]: OK. If `buildable: false` is removed, will spack still build this package if the version specified in this config file satisfies the requirements of the package being `spack install`ed?

File: NERSC/cori_and_edison/packages.yaml (hunk 1 above)
Comment (bcfriesen) [reply NERSC/cori_and_edison/packages.yaml]: Why? Has Cray's Python been responsible for other problems with Spack builds? If so, could we replace this with one of the Conda Python modules that is already provided on Cori? e.g., `python/2.7-anaconda-5.2` for Python 2, `python3/3.6-anaconda-4.4` for Python 3, etc.

Avoiding building Python is responsible for a large reduction in build time in my experiments with Spack at NERSC. Carrying around 15 different versions of Python built with different versions of Intel, GCC, Cray, etc. does not sound like a desirable outcome.

File: NERSC/cori_and_edison/packages.yaml (hunk 2 above)
Comment (bcfriesen) [reply NERSC/cori_and_edison/packages.yaml]: Agreed. But again, if `buildable: false` is removed here, what is the behavior? Will Spack automatically know that it needs to build a newer version of CMake from scratch if a package requires it? Will packages which are satisfied with `/usr/bin/cmake` avoid compiling CMake?

File: NERSC/cori_and_edison/packages.yaml (hunk 1 above)
Comment (mamelara) [reply NERSC/cori_and_edison/packages.yaml]: So actually, I was testing it out with the python modules and things work fine. So I think we should add the anaconda modules there so at least we can have python3 available or can have a more recent python2 version than what the system offers (even though python2 is set to be deprecated in 2020).

File: NERSC/cori_and_edison/packages.yaml (hunk 2 above)
Comment (mamelara) [reply NERSC/cori_and_edison/packages.yaml]: I believe the behavior is if the version declared in `packages.yaml` satisfies what the root package wants then it will use that. Can you confirm @tgamblin?

Comment (mamelara) [reply NERSC/cori_and_edison/packages.yaml]: So Tiffany suggested we add the cmake modules into the `packages.yaml` could we add them here?

File: NERSC/cori_and_edison/packages.yaml [hunk 4]
```diff
@@ -82,3 +82,79 @@ packages:
...
+        paths:
+             readline@6.3: /usr
+    zlib:
+        buildable: false
+        paths:
+             zlib@1.2.8: /usr
+    python:
+        buildable: false
+        paths:
```
Comment (mamelara) [inline NERSC/cori_and_edison/packages.yaml]: Add python modules

File: NERSC/cori_and_edison/packages.yaml (hunk 1 above)
Comment (bcfriesen) [reply NERSC/cori_and_edison/packages.yaml]: Great - I would like to be able to use as much of what Cray already provides in `/usr/bin` as possible. That has already shown to reduce build times drastically for several packages that I've tested.

Although the Conda Python modules are probably newer, Cray does also provide both Python 2 and 3 in `/usr/bin`; could I add `/usr/bin/python3` to this list too?

```bash
friesen@cori08:~> which python
/usr/bin/python
friesen@cori08:~> python --version
Python 2.7.13
friesen@cori08:~> which python3
/usr/bin/python3
friesen@cori08:~> python3 --version
Python 3.4.6
friesen@cori08:~>
```

**Comment by mamelara** [comment]
Hey Brian - how about something like this:

```
packages:
    all:
        compiler: [intel@18.0.1.163, gcc@7.3.0, cce@8.6.5]
        providers:
            mpi: [mpich, openmpi]
            mkl: [intel-mkl]
            blas: [intel-mkl, cray-libsci]
            scalapack: [intel-mkl, cray-libsci]
            pkgconfig: [pkg-config]
    mpich:
        buildable: false
        modules:
            mpich@3.0.1%gcc: cray-mpich/7.7.4
            mpich@3.0.1%intel: cray-mpich/7.7.4
            mpich@3.0.1%cce: cray-mpich/7.7.4
    intel-mkl:
        buildable: false
        paths:
            intel-mkl@2018.1.163%intel+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%gcc+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%cce+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%intel~ilp64 threads=none: /opt/intel
            intel-mkl@2018.1.163%gcc~ilp64 threads=none: /opt/intel
            intel-mkl@2018.1.163%cce~ilp64 threads=none: /opt/intel
    cray-libsci:
        buildable: false
        modules:
            cray-libsci@18.07.1%gcc: cray-libsci/18.07.1
            cray-libsci@18.07.1%intel: cray-libsci/18.07.1
            cray-libsci@18.07.1%cce: cray-libsci/18.07.1
    boost:
        modules:
            boost@1.67.0: boost/1.67.0
            boost@1.69.0: boost/1.69.0
    hdf5:
        modules:
            hdf5@1.10.2.0+hl+fortran~mpi%gcc: cray-hdf5/1.10.2.0
            hdf5@1.10.2.0+hl+fortran~mpi%intel: cray-hdf5/1.10.2.0
            hdf5@1.10.2.0+hl+fortran~mpi%cce: cray-hdf5/1.10.2.0
            hdf5@1.10.2.0+hl+fortran+mpi%gcc: cray-hdf5-parallel/1.10.2.0
            hdf5@1.10.2.0+hl+fortran+mpi%intel: cray-hdf5-parallel/1.10.2.0
            hdf5@1.10.2.0+hl+fortran+mpi%cce: cray-hdf5-parallel/1.10.2.0
    openssl:
        paths:
             openssl@1.0.2j-fips: /usr
    autoconf:
        paths:
             autoconf@2.69: /usr
    automake:
        paths:
             automake@1.13.4: /usr
    bzip2:
        paths:
             bzip2@1.0.6: /usr
    cmake:
        buildable: False
        modules:
            cmake@3.3.2: cmake/3.3.2
            cmake@3.5.2: cmake/3.5.2
            cmake@3.8.2: cmake/3.8.2
            cmake@3.11.4: cmake/3.11.4
            cmake@3.14.0: cmake/3.14.0
    curl:
        paths:
             curl@7.37.0: /usr
    diffutils:
        paths:
             diffutils@3.3: /usr
    gawk:
        paths:
             gawk@4.1.0: /usr
    gettext:
        paths:
             gettext@0.19.2: /usr
    libtool:
        paths:
             libtool@2.4.2: /usr
    lz4:
        paths:
             lz4@1.7.4: /usr
    m4:
        paths:
             m4@1.4.16: /usr
    perl:
        paths:
             perl@5.18.2: /usr
    pkg-config:
        paths:
             pkg-config@0.28: /usr
    tar:
        paths:
             tar@1.27.1: /bin
    xz:
        paths:
             xz@5.0.5: /usr
    readline:
        paths:
             readline@6.3: /usr
    zlib:
        paths:
             zlib@1.2.8: /usr
    python:
        modules:
            python@2.7.13: python/2.7-anaconda-4.4
            python@3.6.8: python/3.6-anaconda-4.4
```

**Comment by bcfriesen** [comment]
> Hey Brian - how about something like this:

Ah I pushed some new commits that do pretty much exactly what you said. How do those look?

**Comment by bcfriesen** [comment]
Preliminary tests of the latest commits I pushed still look good. Spack is picking up the newer CMake and Python modules:
```console
==> cmake@3.14.0 : has external module in cmake/3.14.0
==> cmake@3.14.0 : is actually installed in /usr/common/software/cmake/3.14.0/
==> cmake@3.14.0 : generating module file
==> cmake@3.14.0 : registering into DB
==> mpich@7.7.3 : has external module in cray-mpich
==> mpich@7.7.3 : is actually installed in /opt/cray/pe/mpt/7.7.3/gni/mpich-intel/16.0
==> mpich@7.7.3 : generating module file
==> mpich@7.7.3 : registering into DB
==> python@3.6.8 : has external module in python/3.6-anaconda-5.2
==> python@3.6.8 : is actually installed in /global/common/cori/software/python/3.6-anaconda-5.2
==> python@3.6.8 : generating module file
==> python@3.6.8 : registering into DB
==> Installing amrex
==> Searching for binary cache of amrex
==> Warning: No Spack mirrors are currently configured
==> No binary for amrex found: installing from source
==> Using cached archive: /global/cscratch1/sd/friesen/spack/cache/amrex/amrex-18.10.1.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Staging archive: /tmp/spack/var/spack/stage/amrex-18.10.1-7fudbnszon45xgdiubwy3zodknwqijvi/amrex-18.10.1.tar.gz
==> Created stage in /tmp/spack/var/spack/stage/amrex-18.10.1-7fudbnszon45xgdiubwy3zodknwqijvi
==> No patches needed for amrex
==> Building amrex [CMakePackage]
==> Executing phase: 'cmake'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed amrex
  Fetch: 2.77s.  Build: 3m 8.95s.  Total: 3m 11.72s.
```

**Comment by mamelara** [comment]
Can you add what I added to intel-mkl in my example

```
    intel-mkl:
        buildable: false
        paths:
            intel-mkl@2018.1.163%intel+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%gcc+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%cce+ilp64 threads=tbb: /opt/intel
            intel-mkl@2018.1.163%intel~ilp64 threads=none: /opt/intel
            intel-mkl@2018.1.163%gcc~ilp64 threads=none: /opt/intel
            intel-mkl@2018.1.163%cce~ilp64 threads=none: /opt/intel
```

**Comment by bcfriesen** [comment]
> Can you add what I added to intel-mkl in my example
> 
> ```
>     intel-mkl:
>         buildable: false
>         paths:
>             intel-mkl@2018.1.163%intel+ilp64 threads=tbb: /opt/intel
>             intel-mkl@2018.1.163%gcc+ilp64 threads=tbb: /opt/intel
>             intel-mkl@2018.1.163%cce+ilp64 threads=tbb: /opt/intel
>             intel-mkl@2018.1.163%intel~ilp64 threads=none: /opt/intel
>             intel-mkl@2018.1.163%gcc~ilp64 threads=none: /opt/intel
>             intel-mkl@2018.1.163%cce~ilp64 threads=none: /opt/intel
> ```

Oops, yes I missed those. While we're at it, there are a few other Cray modules that have newer versions available - I will add those too, one moment ...

**Comment by bcfriesen** [comment]
OK I added the `intel-mkl` and included the other versions of the other Cray module already in `packages.yaml`. There are a lot of them, but the Cray module set on the OS image changes only every several months, so they won't need to be updated in `packages.yaml` all that often.

**Comment by bcfriesen** [comment]
So when I add the new Cray modules, the `amrex` package automatically picks up the newest `cray-mpich` module:
```console
==> cmake@3.14.0 : has external module in cmake/3.14.0
==> cmake@3.14.0 : is actually installed in /usr/common/software/cmake/3.14.0/
==> cmake@3.14.0 : generating module file
==> cmake@3.14.0 : registering into DB
==> mpich@7.7.6 : has external module in cray-mpich/7.7.6
==> mpich@7.7.6 : is actually installed in /opt/cray/pe/mpt/7.7.6/gni/mpich-intel/16.0
==> mpich@7.7.6 : generating module file
==> mpich@7.7.6 : registering into DB
==> python@3.6.8 : has external module in python/3.6-anaconda-5.2
==> python@3.6.8 : is actually installed in /global/common/cori/software/python/3.6-anaconda-5.2
==> python@3.6.8 : generating module file
==> python@3.6.8 : registering into DB
==> Installing amrex
==> Searching for binary cache of amrex
==> Warning: No Spack mirrors are currently configured
==> No binary for amrex found: installing from source
==> Using cached archive: /global/cscratch1/sd/friesen/spack/cache/amrex/amrex-18.10.1.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Staging archive: /tmp/spack/var/spack/stage/amrex-18.10.1-s4ahydc3zc5cq7raeg4em2pmv54gsczc/amrex-18.10.1.tar.gz
==> Created stage in /tmp/spack/var/spack/stage/amrex-18.10.1-s4ahydc3zc5cq7raeg4em2pmv54gsczc
==> No patches needed for amrex
==> Building amrex [CMakePackage]
==> Executing phase: 'cmake'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed amrex
  Fetch: 0.00s.  Build: 2m 32.51s.  Total: 2m 32.52s.
```
Is that the expected behavior? It's not a problem, just wondering if Spack will always try to pick the dependency with the highest version if a particular version is not specified.

**Comment by mamelara** [comment]
Yes, that's the expected behavior. I'll merge since it's working for now but I'm going to add a later PR to add some changes that I would like you to test. Thanks again and sorry for delaying on this!
